src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -69,14 +90,50 @@ class PeerManagerActor(
     peer
   }
 
+  def getPeers(): Future[Peers] = {
+    implicit val timeout = Timeout(2.seconds)
+
+    Future.traverse(peers.values) { peer =>
+      (peer.ref ? PeerActor.GetStatus)
+        .mapTo[PeerActor.StatusResponse]
+        .map(sr => (peer, sr.status))
+    }.map(r => Peers.apply(r.toMap))
+  }
+
+  def tryDiscardPeersToFreeUpLimit(): Future[Unit] = {
+    getPeers() flatMap { peers =>
+      val peersToPotentiallyDisconnect = peers.peers.filter {
+        case (_, Handshaking(numRetries)) if numRetries > 0 => true /* still handshaking and retried at least once */
+        case (_, PeerActor.Status.Disconnected) => true /* already disconnected */
+        case _ => false
+      }
+
+      if (peers.peers.size - peersToPotentiallyDisconnect.size >= peerConfiguration.maxPeers) {
+        Future.failed(new RuntimeException("Too many peers"))
```

Comment:
This also happened to me while `max-peers=3`, could that be the case? I think it has something to do with the fact that many of the peers have a `Connecting` status and this is not considered in the `peersToPotentiallyDisconnect` peers.